Library/ViewModels/ReportProjectFormViewModelTests.swift

```diff
@@ -1,43 +1,65 @@
 @testable import KsApi
 @testable import Library
-import ReactiveExtensions
-import ReactiveExtensions_TestHelpers
-import ReactiveSwift
+
 import XCTest
 
 final class ReportProjectFormViewModelTests: TestCase {
-  fileprivate let vm: ReportProjectFormViewModelType = ReportProjectFormViewModel()
-
-  private let userEmail = TestObserver<String, Never>()
-
   // MARK: Computed Properties
 
   private var userEmailSuccessMockService: MockService {
-    let fetchUserQueryData = GraphAPI.FetchUserQuery
+    let fetchUserEmailQueryData = GraphAPI.FetchUserEmailQuery
       .Data(unsafeResultMap: GraphUserEnvelopeTemplates.userJSONDict)
-    guard let envelope = UserEnvelope<GraphUser>.userEnvelope(from: fetchUserQueryData) else {
+
+    guard let envelope = UserEnvelope<GraphUserEmail>.userEnvelope(from: fetchUserEmailQueryData) else {
       return MockService()
     }
 
     let mockService = MockService(
-      fetchGraphUserResult: .success(envelope)
+      fetchGraphUserEmailResult: .success(envelope)
     )
 
     return mockService
   }
 
-  override func setUp() {
-    super.setUp()
-
-    self.vm.outputs.userEmail.observe(self.userEmail.observer)
-  }
-
   func testEmailText_AfterFetchingUsersEmail() {
+    let vm = ReportProjectFormViewModel(
+      projectID: "123", projectURL: "https://www.kickstarter.com/projects/foo/bar",
+      projectFlaggingKind: GraphAPI.FlaggingKind.commentDoxxing
+    )
+
     withEnvironment(apiService: self.userEmailSuccessMockService) {
-      self.vm.inputs.viewDidLoad()
+      let userEmail = CombineTestObserver<String?, Never>()
+      userEmail.observe(vm.$retrievedEmail)
+
+      XCTAssertEqual(userEmail.events.count, 1)
+      XCTAssertEqual(userEmail.events.last, Optional.some(nil))
+
+      vm.inputs.viewDidLoad()
       self.scheduler.advance()
 
-      self.userEmail.assertValues(["nativesquad@ksr.com"])
+      XCTAssertEqual(userEmail.events.count, 2)
+      XCTAssertEqual(userEmail.events.last, Optional("nativesquad@ksr.com"))
     }
   }
+
+  func test_submitIsDisabled_untilDetailTextIsNotEmpty() {
+    let vm = ReportProjectFormViewModel(
+      projectID: "123", projectURL: "https://www.kickstarter.com/projects/foo/bar",
+      projectFlaggingKind: GraphAPI.FlaggingKind.commentDoxxing
+    )
+
+    let saveButtonEnabled = CombineTestObserver<Bool, Never>()
+    saveButtonEnabled.observe(vm.$saveButtonEnabled)
+
+    XCTAssertEqual(saveButtonEnabled.events.count, 1)
+    XCTAssertEqual(saveButtonEnabled.events.last, false)
+
+    vm.detailsText = "This is my report. I don't like this project very much."
+    XCTAssertEqual(saveButtonEnabled.events.count, 2)
+    XCTAssertEqual(saveButtonEnabled.events.last, true)
+
+    vm.detailsText = ""
+    XCTAssertEqual(saveButtonEnabled.events.count, 3)
+    XCTAssertEqual(saveButtonEnabled.events.last, false)
```

Comment:
Since `saveButtonEnabled.events` is an array, we _could_ write these assertions in one line. Def not necessary

```suggestion
    let saveButtonEnabled = CombineTestObserver<Bool, Never>()
    saveButtonEnabled.observe(vm.$saveButtonEnabled)

    XCTAssertEqual(saveButtonEnabled.events, [false])

    vm.detailsText = "This is my report. I don't like this project very much."
    XCTAssertEqual(saveButtonEnabled.events, [false, true])

    vm.detailsText = ""
    XCTAssertEqual(saveButtonEnabled.events, [false, true, false])
```